opentelemetry/api/trace/__init__.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The OpenTelemetry tracing API describes the classes used to generate
+distributed traces.
+
+The :class:`.Tracer` class controls access to the execution context, and
+manages span creation. Each operation in a trace is represented by a
+:class:`.Span`, which records the start, end time, and metadata associated with
+the operation.
+
+This module provides abstract (i.e. unimplemented) classes required for
+tracing, and a concrete no-op ``BlankSpan`` that allows applications to use the
+API package alone without a supporting implementation.
+
+The tracer supports implicit and explicit context propagation. By default spans
+are created as children of the currently active span, and the newly-created
+span becomes the new active span::
+
+    from opentelemetry.sdk.trace import tracer
+
+    # Create a new root span, set it as the current span in context
+    with tracer.span("parent"):
+        # Attach a new child and update the current span
+        with tracer.span("child"):
+            do_work():
+        # Close child span, set parent as current
+    # Close parent span, set default span as current
+
+Under explicit context propagation there is no concept of an active span, and
+the caller is responsible for managing the span's lifetime::
+
+    from opentelemetry.sdk.trace import tracer
+    from your.integration import deserialize_span
+
+    parent = deserialize_span(serialized_span)
+    # Explicit parent span assignment
+    with tracer.span("child", parent=parent) as child:
+        do_work(span=child)
+
+Applications should generally use a single global tracer, and use either
+implicit or explicit context propagation consistently throughout.
+
+.. versionadded:: 0.1.0
```

Comment:
This is the [directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded) that the python docs themselves use. We probably don't want to include this for modules/classes in the first release, but should add it for future API changes.

It looks like this in the generated docs:

![Screen Shot 2019-06-03 at 5 23 01 PM](https://user-images.githubusercontent.com/43799105/58842744-7b794300-8624-11e9-8c1b-d365f88dbae5.png)

Reply:
It might not be bad to include that in the first release too: In my experience, forgetting to add such an annotation is all too easy, and if the first release doesn't have them, it looks like the function was already available back then, even though you just forgot to add the directive.

Reply:
That sounds fine to me, but it does break with the convention in the python docs. Would you expect to annotate every module/class/function? Another option is to annotate the module only at first, then add annotations for classes and functions that are added/changed later.

Reply:
Sounds good. I agree that every function would be too much at least.